Fix problems with EffSendTitle

oskarkk opened the pull request:
### Description

```
command /test:
  trigger:
    send title "test" to player for 1 second with fadein {_x}
```

The above script would give you `Severe Error` because `Something went horribly wrong with Skript.` and among other things would print the stack trace. I've changed the `send title` effect so that if any of the time values was `<none>` or anything other than timespan it will just use the default vanilla values (fadein 10, stay 70, fadeout 20, in ticks). That's the way it works in the other effects with time values like [ignite](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/effects/EffIgnite.java), [poison](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/effects/EffPoison.java) or [potion](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/effects/EffPotion.java).

---
**Target Minecraft Versions:** any
**Requirements:** none

TPGamesNL commented on src/main/java/ch/njol/skript/effects/EffSendTitle.java:
```diff
@@ -57,6 +57,10 @@
 					"send subtitle %string% [to %players%]");
 		
 	}
+
+	private final static int DEFAULT_FADEIN = 10;
+	private final static int DEFAULT_STAY = 70;
+	private final static int DEFAULT_FADEOUT = 20;
```
Can you put these above the static block?

oskarkk replied:
Turned out they aren't needed, because passing -1 to `sendTitle` method sets the times to the defaults.

TPGamesNL commented on src/main/java/ch/njol/skript/effects/EffSendTitle.java:
```diff
@@ -88,9 +92,23 @@ protected void execute(final Event e) {
-			out = fadeOut != null ? (int) fadeOut.getSingle(e).getTicks_i() : -1;
+			int in, stay, out;
+
+			if (fadeIn != null && fadeIn.getSingle(e) != null) {
+				in = (int) fadeIn.getSingle(e).getTicks_i();
+			} else {
+				in = DEFAULT_FADEIN;
+			}
+			if (this.stay != null && this.stay.getSingle(e) != null) {
+				stay = (int) this.stay.getSingle(e).getTicks_i();
+			} else {
+				stay = DEFAULT_STAY;
+			}
+			if (fadeOut != null && fadeOut.getSingle(e) != null) {
+				out = (int) fadeOut.getSingle(e).getTicks_i();
+			} else {
+				out = DEFAULT_FADEOUT;
+			}
```
This will call getSingle twice for 1 execution (for each property), which is unwanted behaviour

oskarkk replied:
Fixed

oskarkk commented:
I fixed the issues, but I want to change the scope of this PR - right now the second syntax doesn't work at all.

`send subtitle %text% [to %players%] [for %time span%] [with fade[(-| )]in %time span%] [(and|with) fade[(-| )]out %time span%]`

By the way, what is the difference between `%-string%` and just `%string%` in `registerEffect`? One is optional, and the other is required?

TPGamesNL commented:
Adding a - prefix to the type will make it not be auto filled (e.g. when it's in an [optional] block and the user doesn't specify it) with e.g. an default expression, such as an event value (though IIRC string doesn't have a default expression attached, so it just prevents an error)

oskarkk commented:
So, after a lengthy investigation...

Effects of `/title @p subtitle "subtest"`, when...
- there's no title displayed at the moment:
  - chosen subtitle is not shown but will be shown with the next title (e.g. `/title @p title ""` to show the subtitle)
- a title/subtitle is displayed:
  - title isn't changed, subtitle changes to the chosen text

Effects of `send subtitle "subtest" to player` in Skript 2.6.1, when...
- there's no title displayed at the moment:
  - something strange happens, it's like there's a new invisible title/subtitle, and the chosen subtitle will be really shown only if you change the title while this invisible title/subtitle exists, so you must do `send subtitle "subtest" to player` and then `send title " " to player`, but if you do that backwards, you'll see nothing
- a title/subtitle is displayed: 
  - title/subtitle that was displayed disappears, and then the same as above: nothing unless you set a new title while the invisible title is existing


So, IMO it's broken right now. We can fix it in two ways:

1. Make it behaving exactly like `/title @p subtitle`

2. Make it more logical, so that `send subtitle "subtest" to player` would be eqivalent to `send title " " with subtitle "subtest" to player` (then it would actually _send_ the subtitle to the player, not only _set_ it - it'd be more in line with the syntax).

With option 2, something like this:

```
send title "test" to player
send subtitle "subtest" to player
```

... would only show the subtitle, no title. With the current Skript, the above code will show nothing, and with option 1 would show title "test" with subtitle "subtest".

With option 1 `send subtitle "subtest" to player` won't display subtitle if a subtitle isn't being displayed right now, or it will be displayed with the next title in the future.

I'm asking you @TPGamesNL and @AyhamAl-Ali for your opinions. Option 1 or 2?

----

And for future reference, in the Bukkit API, any of this does nothing (doesn't send empty title or anything):
```
Player.sendTitle(null, null);
Player.sendTitle("", null);
Player.sendTitle(null, "");
Player.sendTitle("", "");
```

So in Skript something like `send title "" to player` would be useless (won't show empty title, if some title was present, that title would still be displayed).

TPGamesNL commented:
IMO it'd be better to stick with what both Minecraft and CB chose for this, which is to not display the subtitle when no title is displayed (option 1). We could add a note to the docs saying that only sending a subtitle won't show anything unless the player already has a title on screen.

oskarkk commented:
It would be nice if we had a way to send a subtitle with an empty title shorter than `send title " " with subtitle "subtest" to player`. Maybe if we could allow passing string with a newline to the title, with the second line acting as the subtitle? So you could do `send title "%nl%subtest" to player`. That's also not the way it works in vanilla commands or bukkit, see:

![image](https://user-images.githubusercontent.com/16339637/161391058-720c15df-4279-4bfb-8ded-512d66cdf467.png)

But adding this behavior to Skript won't hurt in any way.

This:

`send title "%nl%some text" to player for 3 seconds with fadein 1 second and fadeout 1 second`

... is more than a dozen chars shorter than this:

`send title " " with subtitle "some text" to player for 3 seconds with fadein 1 second and fadeout 1 second`

... or these two lines:

`send title " " to player for 3 seconds with fadein 1 second and fadeout 1 second`
`send subtitle "some text" to player`

TPGamesNL commented:
I don't think there's a good reason to make exceptions for new lines in the strings, I think it would only become more confusing to users.

oskarkk commented:
Okay, I've fixed the behavior of "send subtitle" to match `/title @p subtitle` command, and added more info to the doc. This can be merged now, if everything's okay.

> I don't think there's a good reason to make exceptions for new lines in the strings, I think it would only become more confusing to users.

@TPGamesNL I don't see how this could confuse anyone, the subtitle is literally in the new line under the title, and there's no other use for newlines in this effect. It would be natural, `send title "first line%nl%second line"` would be like `message "first line%nl%second line"`.

TPGamesNL commented:
> I don't see how this could confuse anyone, the subtitle is literally in the new line under the title, and there's no other use for newlines in this effect. It would be natural, `send title "first line%nl%second line"` would be like `message "first line%nl%second line"`.

Because nowhere else in Minecraft (vanilla commands, Bukkit API) is the subtitle considered the next line of the title. This is also seen by the fact that the subtitle is smaller, and that actually trying to use a newline character will not do that. Therefore, I don't see why users would think that adding a newline character would work, there's no precedent for them to think so. 
Besides, the nature of Skript is to be similar to English, and IMO requiring a new line expression for a subtitle doesn't fit this, I don't think it's natural.
This also gives users no way to modify subtitles after being sent (sending a subtitle without a title, therefore leaving the title to what it currently is), at least in terms of syntax (an empty string may work, I'm not sure, but at least looking at the syntax it doesn't look like it will keep the title intact).

oskarkk commented:
> Therefore, I don't see why users would think that adding a newline character would work, there's no precedent for them to think so. Besides, the nature of Skript is to be similar to English, and IMO requiring a new line expression for a subtitle doesn't fit this, I don't think it's natural.

I haven't meant to replace the current syntax. Using a newline in title would be only an option, not a "requirement". If you wanted to send a subtitle, you would have a _choice_ between using longer `send title " " with subtitle "some text"` and `send title "%nl%some text"`. The existing code would work the same as ever, nothing would break (because no one uses newlines in title strings). 

> This also gives users no way to modify subtitles after being sent (sending a subtitle without a title, therefore leaving the title to what it currently is), at least in terms of syntax (an empty string may work, I'm not sure, but at least looking at the syntax it doesn't look like it will keep the title intact).

With what I had in mind, `send subtitle "some text" to player` would still work like `/title @p subtitle "some text"`. Really the only thing added to this effect's code would be something like (pseudocode):

```
if title contains "\n":
    lines = title.split("\n")
    title = lines[0]
    subtitle = lines[1]
```

If you wouldn't put a newline in your title string - you wouldn't notice any change in the behavior of this effect.

TPGamesNL commented:
I don't see the benefit of it, besides having a few characters less, but at the cost of being less intuitive I don't like it. But I'd like feedback from other devs

AyhamAl-Ali commented:
IMO newline expression support here will be confusing as TP stated, that behavior is not normal even tho it could be expected by some users

oskarkk commented on src/main/java/ch/njol/skript/effects/EffSendTitle.java:
```diff
@@ -84,19 +87,38 @@ public boolean init(final Expression<?>[] exprs, final int matchedPattern, final
 	protected void execute(final Event e) {
-		String title = this.title != null ? this.title.getSingle(e) : "",
-		sub = subtitle != null ? subtitle.getSingle(e) : null;
+		String tit = title != null ? title.getSingle(e) : null;
```
Before my changes, the names of variables were

(`Expression` -> `int`/`String`)
`title` -> `title`
`subtitle` -> `sub`
`fadeIn` -> `in`
`stay` -> `stay`
`fadeOut` -> `out`

After my changes it is:

`title` -> `tit`
`subtitle` -> `sub`
`fadeIn` -> `in`
`stay` -> `st`
`fadeOut` -> `out`

So, I tried to make it consistent and that's why I've given shortened names to all of these variables, not just some of them. I think it's more clearer that way. Everyone looking at the code should know what "st" is, because its scope is only the next 20 lines. But maybe something like exprTitle, exprSubtitle, exprFadeIn etc. would be better for `Expression` vars, and then int/String vars would be title, subtitle, fadeIn etc.

TPGamesNL replied:
You don't need different names for fields and variables

AyhamAl-Ali replied:
Also, shortening more often used fields/variables should stay meaningful, `in` by itself is a preposition and meaningful enough in its context but `st` isn't, same goes for `tit`

oskarkk replied:
@TPGamesNL So maybe I'll name them all the same?

oskarkk replied:
Ok, so I named the strings/ints the same as expressions  and it looks nice.

oskarkk commented on src/main/java/ch/njol/skript/effects/EffSendTitle.java:
```diff
@@ -34,9 +34,13 @@
+		"use the <a href='effects.html#EffResetTitle'>reset title</a> effect to restore the default values."})
 @Examples({"send title \"Competition Started\" with subtitle \"Have fun, Stay safe!\" to player for 5 seconds",
 		"send title \"Hi %player%\" to player", "send title \"Loot Drop\" with subtitle \"starts in 3 minutes\" to all players",
 		"send title \"Hello %player%!\" with subtitle \"Welcome to our server\" to player for 5 seconds with fadein 1 second and fade out 1 second",
```
Done.

AyhamAl-Ali replied:
First part is done, second is still, see L45 above

```java
"send title \"Hi %player%\" to player", "send title \"Loot Drop\" with subtitle \"starts in 3 minutes\" to all players",
```
is 2 example lines in 1 line, split them to be 
```java
"send title \"Hi %player%\" to player",
"send title \"Loot Drop\" with subtitle \"starts in 3 minutes\" to all players",
```

oskarkk commented on src/main/java/ch/njol/skript/effects/EffSendTitle.java:
```diff
@@ -34,9 +34,13 @@
+		"be sent with the next title. To show only the subtitle, use: <code>send title \" \" with subtitle \"yourtexthere\" to player</code>.",
+		"",
+		"Note: if no input is given for the times, it will keep the ones from the last title sent, " +
+		"use the <a href='effects.html#EffResetTitle'>reset title</a> effect to restore the default values."})
```
Done.

oskarkk commented on src/main/java/ch/njol/skript/effects/EffSendTitle.java:
```diff
@@ -84,19 +87,38 @@ public boolean init(final Expression<?>[] exprs, final int matchedPattern, final
+				in = t != null ? (int) t.getTicks_i() : -1;
+			} else {
+				in = -1;
+			}
```
Done.

AyhamAl-Ali commented on src/main/java/ch/njol/skript/effects/EffSendTitle.java:
```diff
@@ -34,13 +34,21 @@
-		"send title \"Hello %player%!\" with subtitle \"Welcome to our server\" to player for 5 seconds with fadein 1 second and fade out 1 second",
-		"send subtitle \"Party!\" to all players"})
+@Description({
+	"Sends a title/subtitle to the given player(s) with optional fadein/stay/fadeout times for Minecraft versions 1.11 and above. " +
```
Little thing left unchanged, you probably mean a new empty line here so you need to use a comma 
```suggestion
	"Sends a title/subtitle to the given player(s) with optional fadein/stay/fadeout times for Minecraft versions 1.11 and above. ",
```

Once done just mark it as **resolevd**

oskarkk commented:
So, can this be merged now?

TPGamesNL commented:
You can re-request review from people when you handled their requested changes

TheLimeGlass commented:
Can we also include hex colour support for the titles in this pull request? Examples of how to grab `"<##ffffff>white hex colours example"` can be found at https://github.com/SkriptLang/Skript/blob/a45b189ca5bdc601a2dc6e99c74ea429f8c666a4/src/main/java/ch/njol/skript/effects/EffActionBar.java#L68 and my own https://github.com/TheLimeGlass/Skore/blob/de4d186cce945230387fcd46568d958680c9de39/src/main/java/me/limeglass/skore/elements/expressions/ExprTitle.java#L44

AyhamAl-Ali reviewed (changes requested):
LGTM so far \:)

AyhamAl-Ali commented on src/main/java/ch/njol/skript/effects/EffSendTitle.java:
```diff
@@ -34,13 +34,21 @@
+})
+@Examples({
+	"send title \"Competition Started\" with subtitle \"Have fun, Stay safe!\" to player for 5 seconds",
+	"send title \"Hi %player%\" to player", "send title \"Loot Drop\" with subtitle \"starts in 3 minutes\" to all players",
```
One example per line \:)
```suggestion
	"send title \"Hi %player%\" to player",
	"send title \"Loot Drop\" with subtitle \"starts in 3 minutes\" to all players",
```

TPGamesNL commented:
> Can we also include hex colour support for the titles in this pull request? 

This is out of the scope of this PR
